### DISCUSSION TITLE
Does Antlr4 support multiple token types in type()? If not, why not?

### DISCUSSION THREAD

**Discussion by matthew-dean** [antlr/antlr4#4219] [Q&A]
The documentation is surprisingly thin for Antlr4, so I'm not sure if this is possible or not, but I've, in the past, used Chevrotain, which allows you to map tokens to other categories. The idea is that for the parsing phase, you can reference a "category" as a token. It looks like Antlr's equivalent of this is `-> type( x )` but it looks like [you can only map to one type](https://github.com/antlr/antlr4/blob/master/doc/lexer-rules.md#type). Is it possible to do more such that input can be mapped to multiple token types?

**Comment by kaby76** [comment]
In Antlr, there is no notion of multiple types for a token. A token type is an integer, and it is generally immutable when set.

In an Antlr grammar, the name of the lexer rule represents the token type, which you can use in a parser rule, e.g., `PACKAGE` in [packageDeclaration : annotation* PACKAGE qualifiedName ';' ;](https://github.com/antlr/grammars-v4/blob/master/java/java/JavaParser.g4#L41-L43). However, sometimes one might want to use a different name of the token than from the rule name. That's where `-> type()` might come in handy. All this is consistent with what we've all been taught about formal languages, compiler construction, and lexical analysis--for generations.

The Chevrotain documentation does not contain a single discussion of the "categories" field of a token (under [lexing](https://chevrotain.io/docs/tutorial/step1_lexing.html)), and there are only a few closed issues that make reference to the feature. https://github.com/chevrotain/chevrotain/issues/962 https://github.com/Chevrotain/chevrotain/issues/1055 https://github.com/chevrotain/chevrotain/issues/564#issuecomment-349062346. So we can only guess as to how it is supposed to work from one of the eight example grammars.

In the [calculator](https://github.com/Chevrotain/chevrotain/tree/master/examples/grammars/calculator) example, a category is set for [Mult](https://github.com/Chevrotain/chevrotain/blob/6d210a382b6473379aa7e40857ecf42e14b1248f/examples/grammars/calculator/calculator_pure_grammar.js#L40) and for [Div](https://github.com/Chevrotain/chevrotain/blob/6d210a382b6473379aa7e40857ecf42e14b1248f/examples/grammars/calculator/calculator_pure_grammar.js#L45). Where is the field actually used?

* In a call to [CONSUME()](https://github.com/Chevrotain/chevrotain/blob/6d210a382b6473379aa7e40857ecf42e14b1248f/examples/grammars/calculator/calculator_pure_grammar.js#L115). This method is an implementation detail of the recognizer. Antlr grammars are specifications, not implementations.
* In [retrieving a token value](https://github.com/Chevrotain/chevrotain/blob/6d210a382b6473379aa7e40857ecf42e14b1248f/examples/grammars/calculator/calculator_pure_grammar.js#L205).
 
It isn't clear what value a category brings.

**Reply by matthew-dean** [reply]
[Categories in Chevrotain](https://chevrotain.io/docs/features/token_categories.html) are like virtual tokens.

Say you have a lexer spec like:
```
AttributeFlag : [is];
PlainIdent : [a-zA-Z]+;
```
There are places in your _parser_ where you want to accept any identifier, so you might want something like:
```
rule1 : '(' AttributeFlag | PlainIdent ')';
```
However, there may be _many_ places in the parser where you want to accept any identifier, in which case, in Antlr, it looks like you may do something like:
```
AttributeFlag : [is] -> type(Ident);
PlainIdent : [a-zA-Z]+ -> type(Ident);
```
This simplifies the parser:
```
rule1 : '(' Ident ')';
```
...except now we come to a rule where we can _only_ accept "i" or "s"
```
attributeRule: '[' Ident (ws AttributeFlag)? ']'
```

In Chevrotain, you can essentially "alias" tokens to other virtual tokens using categories, and reference them in the parser where any of that token is acceptable, as in:
```
// lexer (using Antlr-like syntax for readability)
Tilde : '~' -> type(Combinator);
Plus : '+' -> type(AdditionOp, Combinator);
Minus : '-' -> type(AdditionOp);

// parser
selector : SelectorToken Combinator SelectorToken;
addition : Number AdditionOp Number;

// or even
language : 'C' Plus Plus; // "Plus" is still a valid token identifier
```

So it allows you to preserve the token type, or use another token type in your parser grammar, which can simplify the parser a lot.

As to this:

> Where is the field actually used? In a call to CONSUME(). This method is an implementation detail of the recognizer.

No, it isn't. It is the specification _and_ the implementation, as in, it is how the specification is written, and can be statically analyzed, generate railroad diagrams, etc. I'm not saying combining the two is the best way to do it (as it _limits_ you to one implementation language), which is one reason why I'm experimenting with Antlr, but each "method" call is also, effectively, a lexer or parser rule definition the same as a `.g4` file.

**Reply by matthew-dean** [reply]
Maybe there's a fundamental piece of Antlr that I'm misunderstanding. If I'm looking at something like [the Java 9 lexer](https://github.com/antlr/grammars-v4/blob/master/java/java9/Java9Lexer.g4), it has a list of keywords, but those keywords aren't referenced in the parser. Is it identifying those those keywords as the keyword but also `Identifier` based on [this](https://github.com/antlr/grammars-v4/blob/master/java/java9/Java9Lexer.g4#L478) such that [this rule](https://github.com/antlr/grammars-v4/blob/master/java/java9/Java9Parser.g4#L1429) matches all keywords?

**Reply by KvanTTT** [reply]
Yes, [identifier](https://github.com/antlr/grammars-v4/blob/master/java/java9/Java9Parser.g4#L1429) parse rule handles all identifiers including contextual keywords. There is a feature request [Token value comparison operator](https://github.com/antlr/antlr4/issues/1965) for keywords handling simplifying, but it's not implemented.

**Reply by kaby76** [reply]
> AttributeFlag : [is];
> PlainIdent : [a-zA-Z]+;
> rule1 : '(' AttributeFlag | PlainIdent ')';
> AttributeFlag : [is] -> type(Ident);
> PlainIdent : [a-zA-Z]+ -> type(Ident);
> rule1 : '(' Ident ')';
> attributeRule: '[' Ident (ws AttributeFlag)? ']'

Referencing AttributeFlag cannot work in Antlr. I checked using both split and combined grammar, and using a `tokens` declared token type (see below example) and a rule that would never match.

Here are the .g4's for a split grammar.
```
lexer grammar TestLexer;
tokens { Ident }
AttributeFlag : [is] -> type(Ident);
PlainIdent : [a-zA-Z]+ -> type(Ident);
WS : [ \t\n\r]+ -> skip;
LP: '(';
RP: ')';
LB: '[';
RB: ']';
S: ';';
```
```
parser grammar TestParser;
options { tokenVocab=TestLexer; }
start : r (';' r)* EOF;
r : rule1 | attributeRule;
rule1 : '(' Ident ')';
attributeRule: '[' Ident AttributeFlag? ']';
```

`-> type(Ident)` means that the token type is `Ident`. The command `antlr4 TestLexer.g4` generates a lexer .tokens file has a mapping of `Ident=1`, and there is no token type `AttributeFlag`. Following the first call, the command `antlr4 TestParser.g4` generates the warning `implicit definition of token AttributeFlag`. While it also generates .tokens file with AttrbiuteFlag, this cannot be returned by any rule in the lexer. The parser fails for input `[abc i]`.
```
$ ./bin/Debug/net7.0/Test.exe -input '( abc ); [ abc]; [abc i]'
line 1:22 extraneous input 'i' expecting {']', AttributeFlag}
```
In other words, a token type is just an integer, not a set of integers.

> If I'm looking at something like [the Java 9 lexer](https://github.com/antlr/grammars-v4/blob/master/java/java9/Java9Lexer.g4), it has a list of keywords, but those keywords aren't referenced in the parser.

This can be confusing, but most keywords are referenced in the parser grammar for java/java9. Antlr creates a table of aliases between lexer symbol name and the string literal it represents. To actually see what lexer rules that are not referenced, the string literals have to be unfolded, which is how Antlr works. Using [Trash](https://github.com/kaby76/Domemtech.Trash), clone the grammars-v4 repo, cd to java/java9, and execute the command `rm -rf xxxx; mkdir xxxx; trparse *.g4 | trfoldlit | trsponge -o xxxx; trparse xxxx/*.g4 | trxgrep 'for $i in (//lexerRuleSpec[not(FRAGMENT)]/TOKEN_REF) return concat($i/text(), " ", count(doc("*")//parserRuleSpec/ruleBlock//TOKEN_REF[text() = $i/text()]))' | grep " 0"`. This command parses the two grammar files, folds the string literals out of the RHS of all parser rules, then counts all occurrences of the lexer symbol on the RHS of a parser rule. The only rules not referenced are GOTO, CONST, and UNDER_SCORE but these are bugs in the grammar. The rests of the rules are referenced. Rules that are skip or hidden should never be referenced in a parser rule.

Yes, `identifier : Identifier | 'to' | 'module' | 'open' | 'with' | 'provides' | 'uses' | 'opens' | 'requires' | 'exports';` is how https://docs.oracle.com/javase/specs/jls/se20/html/jls-3.html#jls-ContextualKeyword is implemented. (Looks like the grammar should be updated for Java 20.)
